libclc/CMakeLists.txt

```diff
@@ -322,22 +322,26 @@ foreach( t ${LIBCLC_TARGETS_TO_BUILD} )
 
     if ( ARCH STREQUAL spirv OR ARCH STREQUAL spirv64 )
       set( build_flags -O0 -finline-hint-functions -DCLC_SPIRV )
-      set( opt_flags )
+      set( clc_opt_flags )
+      # Inline CLC functions into OpenCL builtins
+      set( opencl_opt_flags "--passes=always-inline" )
```

Comment:
I would just run ordinary optimizations and never use the always_inline attribute

Reply:
Do you mean for SPIR-V in the context of this PR (to stop the introduction of new `always_inline` attributes), or generally across the whole of libclc?

I don't mind the former as, yeah, `always_inline` isn't ideal. But the SPIR-V/Mesa targets seem quite explicit that they don't want any optimizations. Limiting the optimizations to just the `always-inline` pass was my way of adhering to that wish as best as I could. I don't know if SPIR-V/Mesa opt out of optimizations because the SPIR-V translator couldn't handle optimized LLVM IR, or produced worse SPIR-V, or crashed, or whether it was just for compile-time reasons. Maybe it doesn't apply any more? I'm happy to do whatever with those targets, as long as someone can vouch for them.

The latter would be a significant change as in practice _every_ function for the AMDGPU/NVPTX targets - OpenCL or CLC - currently has the `always_inline`  attribute set, via `_CLC_DEF`. I'm not sure it's the best time to drop that assumption.

Reply:
I mean across the whole of libclc. 

> But the SPIR-V/Mesa targets seem quite explicit that they don't want any optimizations

This is nonsense. There are optimizations we should restrict as fancy compression, but the libraries should certainly be optimized. 

> out of optimizations because the SPIR-V translator couldn't handle optimized LLVM IR

If anything, I would expect the opposite

Reply:
This is delving back a little while, but certainly one thing we wanted to avoid was inlining. Partly because it generates massive piles of SPIR-V that we have to parse when everything in libclc gets inlined into everything else in libclc, but also partly because it removes our ability to use libclc only for some functionality and not for everything.

Reply:
This from the min/max commit message makes me think we'll be fine:

> Note that clspv and spirv targets don't want to define these functions, so pre-processor macros replace calls to __clc_min with regular min, for example. This means they can use as much of the generic CLC source files
as possible, but where CLC functions would usually call out to an external __clc_min symbol, they call out to an external min symbol. Then they opt out of defining __clc_min itself in their CLC builtins library.

Reply:
Yeah, we in mesa we can map all those SPIR-V CL ExtInst to either a libclc spir-v implementation or something mesa implementation defined.

There is also like the problem with `fma`, which should never get inlined no matter what, because some devices might actually need the emulation, so it's either having two or mor spir-vs (with fma emulation enabled/disabled and potential other options) or one for everything and we just pick whatever we need.

We aren't linking against the libclc SPIR-V on a SPIR-V level, but do it way later, so we have way more flexibility here and can cope with a single variant. However, that relies on builtins not being inlined into other bultins implementations.

Besides that one we are fine with optimizations as long as the SPIR-V remains valid, which if you enable LLVM optimizations isn't simply the case anymore and the SPIRV-LLVM-Translator ends up generating invalid SPIR-V. But this is another issue.

Reply:
I mean... I think it's fine to enable optimizations, just I'd also require the build system to bail if `spirv-val` is unhappy about it. Afaik the Translator also specifically only supports unoptimized LLVM, because it would have to add quite a lot of lowering to make it all work.

However, if it's good enough for libclc, fine, but then please validate the output. It could be good enough, but one thing I know about is that sometimes with optimization enabled we ended up with e.g. 24 bit integers, which are simply invalid SPIR-V for OpenCL (the CL env spec states it needs to be 8, 16, 32 or 64).

Reply:
Thanks for letting me know, that's useful information to consider. I have seen issues with LLVM optimizations generating invalid (or at least unwanted) SPIR-V in the past, so it's good to be reminded of that.

It sounds as though you don't want any inlining, which is fair enough. Inlining builtins would always be a danger if I make this change. I note that in the current libclc library for SPIR-V (before any of my recent refactoring) we have calls to "internal "__clc" functions such as `__clc_tanpi`, `__clc_rootn`, etc. So perhaps it's fine if this and future PRs introduce more of those e.g. `sign` calling `__clc_sign`?

I can undo the change which adds the `always_inline` attribute to CLC functions, and undo the change that runs `opt --passes=always-inline` for SPIR-V. Does that sound okay?

Reply:
yeah, as long as the OpenCL C builtins aren't inlined into the implementations of other builtins (e.g. a builtin implemented using `fma`) we are fine with any internal inlining, just want to keep the calls to OpenCL C builtins there.

And one thing I forgot to mention is, that we rely on the function name mangling to do our custom function lookup/linking, but I guess this won't be touched by any of this here.

Reply:
Thanks, I've updated the PR. Now `sign` calls `__clc_sign` for all types on SPIR-V.

Mangling shouldn't be a problem. OpenCL builtins will maintain their mangling throughout these patches.

Reply:
An alternative good solution would be to have things like `HAVE_HW_FMA32()` compiling to a SPIR-V specialization constant and then we just specialize the SPIR-V, but I'm not sure if it's worth the effort.

But of course things here could also be more formalized for libclc compiled to a generic SPIR-V binary.

There are many possible paths to make it "proper".

But anyway, I'm not in the mood to discuss this any further with this kind of attitude, soo.. have fun.

Reply:
The correct IR implementation of FMA is universally call the llvm.fma intrinsic. There's no need for additional configuration in libclc. The choice to emit a call or whatnot is then a codegen decision, and is subject to nobuiltin-like configuration